### ISSUE TITLE
Userland emulation, API

### ISSUE THREAD

**Issue by LekKit** [LekKit/RVVM#56]
RVVM is usable as a generic CPU execution engine, not just as a complete machine.
This suggests for working userspace emulation, which runs RISC-V code inside a host process, directly accesses it's memory and defers syscall execution to the host kernel.
For convenience, it's also great to expose an API for that, there is a working project that already needs that (https://github.com/X547/UserlandVM), which uses RVVM for HaikuOS RISC-V userland emulation.

Progress tracking:
- [x] Decide on the initial userland API draft (See #55 for discussion on that)
- [x] Implement the Userland API
- [x] Test regressions with rebased UserlandVM as a primary API user
- [ ] Test RVJIT cache tracking
- [ ] Von Neumann RVJIT cache for userland
- [ ] Signal handling
- [x] Upstream ELF loader
- [x] Linux process stack setup
- [x] Simple system call shims
- [ ] Full Linux syscall emulation

Discussions & suggestions are welcome.

**Comment by X547** [comment]
Is it possible to build `librvvm.a` with pkg-config information?

**Comment by X547** [comment]
I tried to adapt UserlandVM to new librvvm API and it currently crash in `riscv_mmu_op`.

**Comment by LekKit** [comment]
> I tried to adapt UserlandVM to new librvvm API and it currently crash in `riscv_mmu_op`.

Can you give a backtrace?

**Comment by LekKit** [comment]
> Is it possible to build `librvvm.a` with pkg-config information?

Will do. I have no experience with pkg-config however 😅

**Comment by X547** [comment]
Some obvious fixes:
``` diff
diff --git a/src/rvvm.c b/src/rvvm.c
index c782080..a3427af 100644
--- a/src/rvvm.c
+++ b/src/rvvm.c
@@ -707,11 +707,15 @@ PUBLIC rvvm_cpu_handle_t rvvm_create_user_thread(rvvm_machine_t* machine)
     vector_emplace_back(machine->harts);
     rvvm_hart_t* vm = &vector_at(machine->harts, vector_size(machine->harts) - 1);
     riscv_hart_init(vm, machine->rv64);
+    vm->machine = machine;
+    vm->mem = machine->mem;
 #ifdef USE_FPU
     // Initialize FPU properly
 #endif
     return (rvvm_cpu_handle_t)vm;
 }
@@ -743,8 +747,9 @@ PUBLIC rvvm_addr_t rvvm_read_cpu_reg(rvvm_cpu_handle_t cpu, size_t reg_id)
         return vm->csr.cause[PRIVILEGE_USER];
     } else if (reg_id == RVVM_REGID_TVAL) {
         return vm->csr.tval[PRIVILEGE_USER];
+    } else {
+        rvvm_warn("Unknown register %d in rvvm_read_cpu_reg()!", (uint32_t)reg_id);
     }
-    rvvm_warn("Unknown register %d in rvvm_read_cpu_reg()!", (uint32_t)reg_id);
     return 0;
 }
 
@@ -763,6 +768,7 @@ PUBLIC void rvvm_write_cpu_reg(rvvm_cpu_handle_t cpu, size_t reg_id, rvvm_addr_t
         vm->csr.cause[PRIVILEGE_USER] = reg;
     } else if (reg_id == RVVM_REGID_TVAL) {
         vm->csr.tval[PRIVILEGE_USER] = reg;
+    } else {
+        rvvm_warn("Unknown register %d in rvvm_write_cpu_reg()!", (uint32_t)reg_id);
     }
-    rvvm_warn("Unknown register %d in rvvm_write_cpu_reg()!", (uint32_t)reg_id);
 }
```

**Comment by LekKit** [comment]
> Some obvious fixes:
> 
> ```diff
> diff --git a/src/rvvm.c b/src/rvvm.c
> index c782080..a3427af 100644
> --- a/src/rvvm.c
> +++ b/src/rvvm.c
> @@ -707,11 +707,15 @@ PUBLIC rvvm_cpu_handle_t rvvm_create_user_thread(rvvm_machine_t* machine)
>      vector_emplace_back(machine->harts);
>      rvvm_hart_t* vm = &vector_at(machine->harts, vector_size(machine->harts) - 1);
>      riscv_hart_init(vm, machine->rv64);
> +    vm->machine = machine;
> +    vm->mem = machine->mem;
>  #ifdef USE_FPU
>      // Initialize FPU properly
>  #endif
>      return (rvvm_cpu_handle_t)vm;
>  }
> @@ -743,8 +747,9 @@ PUBLIC rvvm_addr_t rvvm_read_cpu_reg(rvvm_cpu_handle_t cpu, size_t reg_id)
>          return vm->csr.cause[PRIVILEGE_USER];
>      } else if (reg_id == RVVM_REGID_TVAL) {
>          return vm->csr.tval[PRIVILEGE_USER];
> +    } else {
> +        rvvm_warn("Unknown register %d in rvvm_read_cpu_reg()!", (uint32_t)reg_id);
>      }
> -    rvvm_warn("Unknown register %d in rvvm_read_cpu_reg()!", (uint32_t)reg_id);
>      return 0;
>  }
>  
> @@ -763,6 +768,7 @@ PUBLIC void rvvm_write_cpu_reg(rvvm_cpu_handle_t cpu, size_t reg_id, rvvm_addr_t
>          vm->csr.cause[PRIVILEGE_USER] = reg;
>      } else if (reg_id == RVVM_REGID_TVAL) {
>          vm->csr.tval[PRIVILEGE_USER] = reg;
> +    } else {
> +        rvvm_warn("Unknown register %d in rvvm_write_cpu_reg()!", (uint32_t)reg_id);
>      }
> -    rvvm_warn("Unknown register %d in rvvm_write_cpu_reg()!", (uint32_t)reg_id);
>  }
> ```

Oh... Silly me
Does it work now?

**Comment by LekKit** [comment]
Apart from the possible explosions from RVJIT and my poor checking of my own code, I tried to copy your approach and also used some proper functions for FPU enablement and other cpu state setup

**Comment by X547** [comment]
> Does it work now?

Not yet. Crash after first syscall with NULL guest dereference. Maybe my fault, need to debug more. TinyEMU engine still working.

**Comment by LekKit** [comment]
> > Does it work now?
> 
> Not yet. Crash after first syscall with NULL guest dereference. Maybe my fault, need to debug more. TinyEMU engine still working.

See my comment from d20c2d2 (Concern about PC increment behavior)

> Where should the PC increment happen? In UserlandVM, merely returning from riscv_trap() means we land on a next instruction after the one that triggered the trap. This implementation preserves this behavior, but may need some re-arrangement. For example, we can increment the PC on the next entry to riscv_hart_run_userland(), or perhaps defer this to outside code (Will require changes on UserlandVM side)

This part should be more explicit and well documented. I'm not sure if this is the thing that broke for you, but just be aware.
As for now, I'm not sure how do i set up UserlandVM to debug this myself, and not sure if any intervention is needed. There isn't much code changed so maybe you can wriggle around it and see potential fixes.

**Comment by X547** [comment]
PC register is 4 inside syscall handler.

```
> UserlandVM /boot/home/Tests/UserlandVM/apps/StyledEdit
syscall 193(_kern_create_area)
  PC: 4
[!] unhandled trap: Hart 0x7fda61ed74d0 trap at 00000000, cause 1, tval 0x000004
FP: 0x138f4fc8fd0, PC: 0
FP: 0x138f4fc9000, PC: 0xf6ae739332
FP: 0, PC: 0x209fe946018
```

**Comment by X547** [comment]
Virtual CPU still try to jump at TVEC that is zero. It should be not done for userland emulation mode.

**Comment by LekKit** [comment]
> Virtual CPU still try to jump at TVEC that is zero. It should be not done for userland emulation mode.

Weird. I'm not sure how that's happening (Unless you are calling riscv_hart_run() or clearing vm->user_traps), I grep'ed the source tree for manupulations on REGISTER_PC and did not see any other way.

**Comment by X547** [comment]
I used older version of `riscv_hart.c` 😅.

Now it seems crash on FPU instruction:
```
syscall 223(_kern_write_port_etc)
  PC: 1052861742872
syscall 219(_kern_port_buffer_size_etc)
  PC: 1052861742808
syscall 221(_kern_read_port_etc)
  PC: 1052861742840
[!] unhandled trap: Hart 0x7f314e9fc940 trap at 0x193b9f4cc70, cause 2, tval 0x102773
FP: 0xb4e19de740, PC: </boot/home/Tests/UserlandVM/apps/lib/libbe.so> 0x193c70
FP: 0xb4e19de7e0, PC: </boot/home/Tests/UserlandVM/apps/lib/libtracker.so> 0x11362a
FP: 0xb4e19de820, PC: </boot/home/Tests/UserlandVM/apps/lib/libtracker.so> 0x11392e
FP: 0xb4e19de8c0, PC: </boot/home/Tests/UserlandVM/apps/lib/libtracker.so> 0xe30ce
FP: 0xb4e19dead0, PC: </boot/home/Tests/UserlandVM/apps/lib/libtracker.so> 0xe4738
FP: 0xb4e19dec50, PC: </boot/home/Tests/UserlandVM/apps/lib/libtracker.so> 0xe53ac
FP: 0xb4e19ded20, PC: </boot/home/Tests/UserlandVM/apps/lib/libtracker.so> 0xe0156
FP: 0xb4e19dedf0, PC: </boot/home/Tests/UserlandVM/apps/StyledEdit> 0x152c0
FP: 0xb4e19defa0, PC: </boot/home/Tests/UserlandVM/apps/StyledEdit> 0x12100
FP: 0xb4e19defd0, PC: </boot/home/Tests/UserlandVM/apps/StyledEdit> 0x121f6
FP: 0xb4e19df000, PC: 0x9476af3370
FP: 0, PC: 0x10e62d5e018
```

**Comment by X547** [comment]
It started after recovering my old HART state initialization code:
```
    maxlen_t mstatus = 0xA00000000 + (FS_INITIAL << 13);
    riscv_csr_op(vm, 0x300, &mstatus, CSR_SWAP);
```
![screenshot492](https://user-images.githubusercontent.com/45505339/202718157-a55b9612-8787-4358-9869-1cfbfba17c65.png)

**Comment by LekKit** [comment]
> It started after recovering my old HART state initialization code:

Neat! I will provide upstream fixes soon.
Does RVJIT work? 😅

**Comment by X547** [comment]
Seems yes, but not with all programs.

![screenshot493](https://user-images.githubusercontent.com/45505339/202719508-1fd7b4bd-9aa4-413e-bab7-361c2e56a3f2.png)

**Comment by LekKit** [comment]
> Seems yes, but not with all programs.

Well, considering the absolute lack of cache coherence in userland mode, it's understandable.
I think there are places where you could've missed the need to call rvvm_icache_flush(), for example munmap syscalls and other stuff that invalidates the code in memory.
Alas, rvvm_icache_flush() is also potentially racy. I don't know for sure, would need to have your environment & working ThreadSanitizer. I remember you tried wrapping cpu runner in a lock for TEmu, so maybe you can try the same & see if it fixes anything.

**Comment by X547** [comment]
Something wrong is happening with multiple threads (host heap corruption etc.).

**Comment by X547** [comment]
> This part should be more explicit and well documented. I'm not sure if this is the thing that broke for you, but just be aware.

Real RISC-V behavior is not incrementing PC on ECALL trap. This is also what TinyEMU does: https://github.com/X547/UserlandVM/blob/master/VirtualCpuRiscVTemu.cpp#L48.

**Comment by LekKit** [comment]
> Real RISC-V behavior is not incrementing PC on ECALL trap. This is also what TinyEMU does: https://github.com/X547/UserlandVM/blob/master/VirtualCpuRiscVTemu.cpp#L48.

I know, it's just an implementation detail (And optimization) of RVVM interpreter that leaked when we abandoned machine mode. After each interpreted instruction it increments PC by instruction size, but the trap handling jumps to TVEC which makes the PC increment invisible to the guest.

Omitting the increment for userland emulation also adds confusion when it's not the ECALL instruction: we aren't sure what's the size of the instruction (Which is why it was initially done this way internally)

**Comment by X547** [comment]
`rvvm_create_user_thread` have no `machine->harts` thread protection and `rvvm_free_user_thread` do not remove `hart` from list.

**Comment by LekKit** [comment]
> `rvvm_create_user_thread` have no `machine->harts` thread protection and `rvvm_free_user_thread` do not remove `hart` from list.

There are some disastrous things from that vector manipulation already with CPU hotplug, a patch is on the way. You can try placing ```spin_lock(&global_lock)``` there, but another horrific issue arises from the vector reallocation which moves our cpus in memory (OH GOD).
So heh, i think another round or two of patches is expected.

**Comment by LekKit** [comment]
Just looked at my ```git diff --stat```

```21 files changed, 3502 insertions(+), 3152 deletions(-)```
Oh... Poor networking, hid & fb stuff... I probably should start upstreaming them somehow

**Comment by X547** [comment]
I removed `hart` registration in `machine` and more complex applications started working (no JIT).

![screenshot494](https://user-images.githubusercontent.com/45505339/202732009-6cef118d-4575-40e6-92f8-9ad6967c1cfc.png)

**Comment by X547** [comment]
It also works with JIT. `hart` registration is currently main problem of multi-thread app failures.

**Comment by LekKit** [comment]
> It also works with JIT. `hart` registration is currently main problem of multi-thread app failures.

Hmm, thanks.
I will switch from this task for a bit if you're alright with that, have some more stuff to do (Both related and unrelated to RVVM).
I also have some questions about Haiku networking, and some HID API design decisions which need approval, but I'm not sure if that's related to any of the opened issues.

**Comment by LekKit** [comment]
@X547 Can you make some kind of torture test for this API that will work outside of Haiku? I need that so I can test it using ThreadSanitizer & other stuff.

**Comment by X547** [comment]
I made UserlandVM for Haiku only. Currently I have no tests that will work on Linux. There is a project that implement Haiku syscalls on Linux at library level (`_kern_*` functions in `libroot.so`), maybe it can be used to run something like `7z b` or JavaScript/Lua JIT.

Maybe it is not too hard to make some basic userland VM for Linux, but I am not familiar with Linux syscalls.

Can you fix multiple virtual HART registration/unregistration first? I think it can be done without advanced debugging tools like ThreadSatinizer.

**Comment by LekKit** [comment]
> Can you fix multiple virtual HART registration/unregistration first? I think it can be done without advanced debugging tools like ThreadSatinizer.

Done, see staging branch. Be aware I've been committing code without any testing because I have no idea how to get this working. It compiles & passes static analysis at least, but otherwise you're the only user of this API who can confirm if I did things right.

**Comment by LekKit** [comment]
> There is a project that implement Haiku syscalls on Linux at library level (`_kern_*` functions in `libroot.so`), maybe it can be used to run something like `7z b` or JavaScript/Lua JIT.

Where do I get it? Sounds good

> Maybe it is not too hard to make some basic userland VM for Linux, but I am not familiar with Linux syscalls.

Will be implemented at some point, but right now I'm more focused at upstreaming machine-related stuff (Networking) and helping you with existing userland emulator instead of writing one from scratch.

**Comment by X547** [comment]
Latest version fails to build for me:
```
> git log
commit 3e970a388d333e3933c36437a6cf855b7d9626ff (HEAD -> master3, origin/staging)
Author: LekKit <50500857+LekKit@users.noreply.github.com>
Date:   Sun Nov 20 13:20:32 2022 +0200

    Use machine->harts pointers properly in PLIC

> git diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 749cfbe..faddc34 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -161,9 +161,9 @@ list(REMOVE_ITEM RVVM_DEVICES_SRC
 
 if (RVVM_USE_FB)
        if (HAIKU)
-               target_sources(rvvm PRIVATE "${RVVM_HAIKU_SRC}")
+               target_sources(rvvm_common INTERFACE "${RVVM_HAIKU_SRC}")
                find_library(RVVM_BE_LIB be REQUIRED)
-               target_link_libraries(rvvm PRIVATE ${RVVM_BE_LIB})
+               target_link_libraries(rvvm_common INTERFACE ${RVVM_BE_LIB})
        elseif (UNIX)
                find_package(X11)
                if (NOT TARGET X11::X11)
> cmake .. -DRVVM_USE_FB=ON -DRVVM_USE_JIT=ON -DCMAKE_BUILD_TYPE=Release -G Ninja
> ninja
[0/2] Re-checking globbed directories...
[56/56] Linking CXX executable rvvm
FAILED: rvvm 
: && /bin/c++ -O3 -DNDEBUG -flto -fno-fat-lto-objects -flto=auto CMakeFiles/rvvm_bin.dir/src/main.c.o CMakeFiles/rvvm_bin.dir/src/devices/haiku_window.cpp.o -o rvvm  -Wl,-rpath,/boot/system/lib  librvvm.a  librvvm_cpu32.a  librvvm_cpu64.a  librvjit.a  librvvm_cpu32.a  librvvm_cpu64.a  librvjit.a  /boot/system/lib/libatomic.so  /boot/system/develop/lib/libm.a  /boot/system/lib/libbe.so && :
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: /tmp//ccUAaQQJ.ltrans3.ltrans.o: in function `rvjit_ctx_free':
<artificial>:(.text+0x1ab): undefined reference to `hashmap_clear'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0x1b5): undefined reference to `hashmap_destroy'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0x1bd): undefined reference to `hashmap_destroy'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: /tmp//ccUAaQQJ.ltrans3.ltrans.o: in function `rvjit_block_lookup':
<artificial>:(.text+0x3e1): undefined reference to `hashmap_rebalance'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0x49b): undefined reference to `hashmap_rebalance'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0x4ed): undefined reference to `hashmap_shrink'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0x501): undefined reference to `hashmap_shrink'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: /tmp//ccUAaQQJ.ltrans3.ltrans.o: in function `rvjit_flush_cache':
<artificial>:(.text+0x857): undefined reference to `hashmap_clear'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0x898): undefined reference to `hashmap_clear'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: /tmp//ccUAaQQJ.ltrans3.ltrans.o: in function `rvjit_ctx_init':
<artificial>:(.text+0xed7): undefined reference to `hashmap_init'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0xee5): undefined reference to `hashmap_init'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: /tmp//ccUAaQQJ.ltrans3.ltrans.o: in function `rvjit_block_finalize':
<artificial>:(.text+0x352f): undefined reference to `hashmap_rebalance'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0x3620): undefined reference to `hashmap_grow'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0x3656): undefined reference to `hashmap_shrink'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0x366d): undefined reference to `hashmap_rebalance'
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/11.2.0/../../../../x86_64-unknown-haiku/bin/ld: <artificial>:(.text+0x368e): undefined reference to `hashmap_grow'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by X547** [comment]
But compiled fine with Clang, strange... Some LTO-related trouble?

Confirmed that UserlandVM works with JIT and multiple threads with latest RVVM version.

**Comment by X547** [comment]
> Where do I get it? Sounds good

https://github.com/trungnt2910/hyclone

**Comment by LekKit** [comment]
> But compiled fine with Clang, strange... Some LTO-related trouble?

What's even weirder, I managed to reproduce this on Linux where it surely should not hit that if (HAIKU) branch. This still happens after rolling CMakeLists back, which means either it was broken all this time or I somehow broke it by source tree changes (???). This happens regardless of LTO and only with GCC, not with Clang or MSVC. Reminds me of similar issue in GCC related to inline functions: if you ever define an inline function without static attribute, it will later fail at link stage with optimizations disabled. We sure are not violating this rule though, since otherwise ```make CFLAGS=-O0``` would also fail.

**Comment by LekKit** [comment]
I summoned the CMake Wizard, @cerg2010cerg2010 to help with this. He mentions that his initial logic to separate librvjit from librvvm may have been broken due to the fact we include librvvm parts from there (hashmap, etc)

**Comment by X547** [comment]
> What's even weirder, I managed to reproduce this on Linux where it surely should not hit that if (HAIKU) branch.

It can be link order trouble. GNU `ld` is sensitive to object files link order, but LLVM `lld` isn't.

**Comment by LekKit** [comment]
> Latest version fails to build for me:

Fixed in 7b9b375

**Comment by LekKit** [comment]
This issue remains as a tracker. If any app doesn't work under UserlandVM and you believe it to be RVVM bug, please report here (With additional information, does it break with JIT or completely, etc). Thanks.

The remaining tasks in this issue are postponed to collect more information first.

**Comment by X547** [comment]
I found one regression with new `librvvm`: menu windows are displayed at wrong position and size. Issue is not present with old embedded RVVM code and with TinyEMU code. Probably FPU related because `float` type is used in GUI coordinates.

rvvm:
![screenshot495](https://user-images.githubusercontent.com/45505339/202908174-68ebb82d-c8e7-45d2-ad87-74c146cc7a89.png)

rvvm2 (note grey thin rectangle at screen left side):
![screenshot496](https://user-images.githubusercontent.com/45505339/202908181-c34d75a2-c435-4993-a52f-3694775e5ff8.png)

**Comment by LekKit** [comment]
> Probably FPU related because `float` type is used in GUI coordinates.

You may try rolling back src/cpu/riscv_f.c to older commits, there weren't much breaking changes relative to the other code base. Be aware that it was also slightly broken previously (Explained in each commit).
Notice also, our FPU needs CFLAG -frounding-math to pass RISC-V tests because compiler madness... and I don't see that CMake sets such flag. I also had reports that -frounding-math is unsupported on older Clang.

**Comment by LekKit** [comment]
Public CPU register accessors may also be at fault. They bitcast FPU registers into u64, and either this isn't right or you are misinterpreting that data

**Comment by X547** [comment]
> Public CPU register accessors may also be at fault. They bitcast FPU registers into u64, and either this isn't right or you are misinterpreting that data

I tried to turn off FPU registers access in rvvm2 engine add-on (UserlandVM currently do not touch FPU registers anyway) and it have no effect.

**Comment by LekKit** [comment]
> > Public CPU register accessors may also be at fault. They bitcast FPU registers into u64, and either this isn't right or you are misinterpreting that data
> 
> I tried to turn off FPU registers access in rvvm2 engine add-on (UserlandVM currently do not touch FPU registers anyway) and it have no effect.

Does it work for rvvm1 though? Just asking 😅
I guess this is either a regression (easily bisectable since the FPU is a single source with stable API), a broken build (Compilers sure hell like to break FPU code) or something in your code is messing up FPU env (rounding mode, exceptions). 
It could also be unrelated to FPU, although i'm not sure what changed in order for that to happen...

**Comment by X547** [comment]
> Does it work for rvvm1 though? Just asking 😅

Yes (first screenshot).

**Comment by LekKit** [comment]
> Probably FPU related because `float` type is used in GUI coordinates.

Question: Does the software in question work properly inside Haiku RISC-V guest in RVVM machine instead of userland emulation?

**Comment by X547** [comment]
> Question: Does the software in question work properly inside Haiku RISC-V guest in RVVM machine instead of userland emulation?

Timer interrupts are currently broken in Haiku RVVM guest. I tried various combinations (OpenSBI timers and haiku_loader.riscv machine mode timer control code). Works with TinyEMU and QEmu.

**Comment by LekKit** [comment]
> > Question: Does the software in question work properly inside Haiku RISC-V guest in RVVM machine instead of userland emulation?
> 
> Timer interrupts are currently broken in Haiku RVVM guest. I tried various combinations (OpenSBI timers and haiku_loader.riscv machine mode timer control code). Works with TinyEMU and QEmu.

Can you give images for investigation?

**Comment by X547** [comment]
Can you give a patch that restore ATA driver?

**Comment by LekKit** [comment]
> Can you give a patch that restore ATA driver?

Sure thing

```patch
diff --git a/src/main.c b/src/main.c
index 3d66493..d2043a5 100644
--- a/src/main.c
+++ b/src/main.c
@@ -32,6 +32,8 @@ along with this program.  If not, see <https://www.gnu.org/licenses/>.
 #include "devices/nvme.h"
 #include "devices/eth-oc.h"
 
+#include "devices/ata.h"
+
 #include <stdio.h>
 #include <string.h>
 #include <inttypes.h>
@@ -231,7 +233,8 @@ static int rvvm_main(int argc, const char** argv)
         } else if (cmp_arg(arg_name, "k") || cmp_arg(arg_name, "kernel")) {
             success = success && rvvm_load_kernel(machine, arg_val);
         } else if (cmp_arg(arg_name, "i") || cmp_arg(arg_name, "image")) {
-            success = success && nvme_init_auto(machine, arg_val, true);
+            //success = success && nvme_init_auto(machine, arg_val, true);
+            ata_init_auto(machine, rvvm_get_pci_bus(machine), blk_open(arg_val, BLKDEV_RW));
         } else if (cmp_arg(arg_name, "cmdline")) {
             rvvm_cmdline_set(machine, arg_val);
         } else if (cmp_arg(arg_name, "append")) {
```

I may bring back some kind of ```-ata``` option if you really need it. For now it's limited to librvvm as an API to create such devices.

**Comment by X547** [comment]
```
WARN: Possible deadlock at /Haiku/data/packages/RVVM/src/devices/ns16550a.c@267
WARN: The lock was previously held at /Haiku/data/packages/RVVM/src/devices/ns16550a.c@168
WARN: Attempting to recover execution...
 * * * * * * *
```
Repeatedly displayed.

**Comment by LekKit** [comment]
> ```
> WARN: Possible deadlock at /Haiku/data/packages/RVVM/src/devices/ns16550a.c@267
> WARN: The lock was previously held at /Haiku/data/packages/RVVM/src/devices/ns16550a.c@168
> WARN: Attempting to recover execution...
>  * * * * * * *
> ```

Whoosh. Some CPU thread was stuck inside uart mmio handler for too long, I don't think there are any actual locking issues in the driver. This may be caused by blocking syscalls on stdin/stdout.

Both serial read & write to UART are followed either by select() with timeout 0 to check pending bytes in stdin, or write() to stdout. This could block if you switch them to something that isn't a TTY, or if the OS does janky stuff.

Hmm, another idea. I had similar repeated pseudo-deadlock issues when messing with rvtimer locally (Which caused bogus timeouts and would also directly affect guest timers and you are telling me exactly that...)

**Comment by X547** [comment]
One potential timer bug: https://github.com/LekKit/RVVM/blob/09575a0c6c1d9610cdee127195ce0322461db593/src/riscv_priv.c#L88. Timer pending interrupt is set after lock, but lock can be also released by other interrupts.

**Comment by X547** [comment]
It started with ATA disk, but I can't control system because of missing Altera PS/2 support. Critical timer problems seems fixed (CPU meter changes can be observed).

![screenshot497](https://user-images.githubusercontent.com/45505339/203144761-fb9a472f-a52e-4db8-8279-ddd37901ae67.png)

**Comment by LekKit** [comment]
> One potential timer bug:
> 
> https://github.com/LekKit/RVVM/blob/09575a0c6c1d9610cdee127195ce0322461db593/src/riscv_priv.c#L88
> . Timer pending interrupt is set after lock, but lock can be also released by other interrupts.

You are probably right. Nothing broke for Linux/RV tests so far after fixing this, and this probably falls into "bogus IRQ" category. I will merge this after some more testing.

```c
if (rvtimer_pending(&vm->timer)) vm->csr.ip |= (1 << INTERRUPT_MTIMER);
```

**Comment by LekKit** [comment]
> Missing Altera PS/2 support

If this is the only broken thing for now, then great. Unfortunately it'll take me a while to implement proper input devices, and the **Wizard of the Cursed Devices** (cerg2010) doesn't have enough time for such contributions. Don't bother on Altera support in Haiku, I'm unsure this device physically exists anyways (It's an FPGA).

What's up with uart deadlock reporting? Did it disappear somehow or still reporting garbage?
The only time I've seen this is on a very loaded system in near-death condition, where one thread would get preempted for eternity while holding a lock, so the other thread freaked out.

**Comment by X547** [comment]
> If this is the only broken thing for now, then great.

Seems yes for now, but it is hard to check more when system is uncontrollable (no networking too). Some autostart scripts can be used for testing.

> What's up with uart deadlock reporting?

It only happens in kernel debugger that use UART polling instead of interrupts.

**Comment by LekKit** [comment]
> It only happens in kernel debugger that use UART polling instead of interrupts.

And you are running this on a Haiku host, right?
I'll look into that. The locks are implemented in such a way that they are **usually** fair to contention, but maybe something caused one thread to always claim the lock first, then the other one was never able to claim it (And complained)

**Comment by LekKit** [comment]
> One potential timer bug:

I investigated further, and this is not a bug. I completely forgot what that thing was for, basically setting INTERRUPT_MTIMER is only a hint to interrupt dispatcher "check actual timer value, now" and defers the actual interrupt there.
This is intended because:
- RVVM scheduler thread sends preemptive interrupts (Not the ones handled in WFI), but the hart should then check on it's side whether those interrupts are still pending (We could've handled them already, someone changed timecmp, etc).
- Needing to check timer only in a single place upon hint reduces amount of timer syscalls

Same trick is done inside WFI to improve timer precision, i.e. we wait on the wfi_cond, then hint the dispatch to check timer and immediately go there using riscv_restart_dispatch(). I should document that behaviour properly, because even I forgot what that was for, but that's working properly.

To be fair, even if this did send spurious timer interrupts, this shouldn't have affected the guest for reasons already explained above. Interrupts are asynchronous, could be delivered to the hart with unspecified latency, and basically they are more like a hint "something *probably* happened".

Consider following example of spurious interrupts:
- Kernel submits two IO requests to NVMe submission queue
- NVMe completes first request, writes into completion queue, sends an interrupt
- A hart enters the interrupt handler
- During this brief moment, NVMe completes a second request & writes it's completion
- The kernel drains the completion queue, sees both IO requests are completed
- Only now the interrupt from the second request completion is delivered to the hart due to a *tiny* race between them. The kernel sees completion queue is empty and concludes it was a spurious interrupt.

**Comment by X547** [comment]
> Question: Does the software in question work properly inside Haiku RISC-V guest in RVVM machine instead of userland emulation?

Confirmed that window position problem is not observed in Haiku guest (tested by adding Altera PS/2 support).

**Comment by LekKit** [comment]
> > Question: Does the software in question work properly inside Haiku RISC-V guest in RVVM machine instead of userland emulation?
> 
> Confirmed that window position problem is not observed in Haiku guest (tested by adding Altera PS/2 support).

Woah, you actually added that device :D

Does your user emulator use FPU outside of RVVM code? If yes, it could be screwing over guest FPU exceptions, try fesetenv/fegetenv functions around rvvm_run_user_thread() (I could add this into RVVM in future, but just for sake of quick testing).
Also please check if you build the userland VM properly (already explained here about optimizaton options and stuff). Run RV tests if you can, but they are designed for M-mode (With RVVM syscon & uart for result reporting).
[riscv-tests.tar.gz](https://github.com/LekKit/RVVM/files/10087746/riscv-tests.tar.gz)

Perhaps I should join the debugging efforts, but I need UserlandVM sources which build around staging rvvm apis, and guest binaries. Would be glad if you shared those, in form of separate UserlandVM branch and binaries uploaded somewhere.

**Comment by LekKit** [comment]
Initial Linux userspace emulation efforts
I have some ABI issues with C binaries tho, they are looking somewhere beyond top of the stack and use pointers from there (argv / envp, perhaps?). Hand-made ASM binaries work fine for now.

![image](https://user-images.githubusercontent.com/50500857/204134772-dcba6579-ad5f-447a-80d7-6f3c45fa46ab.png)

**Comment by X547** [comment]
There is also [libriscv](https://github.com/fwsGonzo/libriscv) project with its own emulation engine and it provides Linux syscalls support over POSIX/WinAPI. Maybe it can be used as reference.

**Comment by LekKit** [comment]
> There is also [libriscv](https://github.com/fwsGonzo/libriscv) project with its own emulation engine and it provides Linux syscalls support over POSIX/WinAPI. Maybe it can be used as reference.

I've similar plans for linux syscalls over POSIX at least (May use Cygwin for win32, if not it's better to make a separate POSIX emulation layer project).
For now tho I want something simple to test FPU code in userspace and provide fixes for your UserlandVM use case.

**Comment by X547** [comment]
> Perhaps I should join the debugging efforts, but I need UserlandVM sources which build around staging rvvm apis, and guest binaries. Would be glad if you shared those, in form of separate UserlandVM branch and binaries uploaded somewhere.

I send current UserlandVM data including guest riscv64 haiku binaries. It is not yet well organized and not ready for end user.

Run with command like this:
`~/Tests/UserlandVM/build.x86_64> UserlandVM --engine ./librvvm2.so ../apps/StyledEdit`

`rvvm.pc` (librvvm pkg-config file needed to build UserlandVM):
```
prefix=/Haiku/data/packages/RVVM
includedir=${prefix}/src
libdir=${prefix}/build.x86_64

Name: rvvm
Description: The RISC-V Virtual Machine
Version: 0.5.0
Libs: -L${libdir} -lrvvm_cpu32 -lrvvm_cpu64 -lrvjit  -lrvvm
Cflags: -I${includedir}
```

[UserlandVM.7z.zip](https://github.com/LekKit/RVVM/files/10098327/UserlandVM.7z.zip)

**Comment by X547** [comment]
Please tell if you managed to run UserlandVM or have some troubles with it.

**Comment by LekKit** [comment]
> Please tell if you managed to run UserlandVM or have some troubles with it.

All CPU engines catch unhandled trap. Both legacy and rvvm2, and even Temu (Tho Temu has a trap cause 1).
Both are NULL deref, tho in case of RVVM it's a load and Temu has instruction fetch fault.
I didn't change anything yet, didn't rebuild. Running on hrev56695 x86_64.

![image](https://user-images.githubusercontent.com/50500857/204142295-047b9d68-55e9-4b21-926f-2977371ec9ee.png)

**Comment by X547** [comment]
> All CPU engines catch unhandled trap.

Probably app_server protocol breakage in https://github.com/haiku/haiku/commit/241f109ccbc17646932772fb9bd5b9e7c99d5d85, so need to rebuild guest `libbe.so`.

**Comment by LekKit** [comment]
> > All CPU engines catch unhandled trap.
> 
> Probably app_server protocol breakage in [haiku/haiku@241f109](https://github.com/haiku/haiku/commit/241f109ccbc17646932772fb9bd5b9e7c99d5d85), so need to rebuild guest `libbe.so`.

Will updating the host help?

UDP: It didn't, but Temu now actually catches SEGV in get_insn32()

**Comment by X547** [comment]
> Running on hrev56695

There are no such revision yet. app_server protocol is changed in hrev56593, so downgrading to older version should help.

**Comment by X547** [comment]
It seems actually a different problem (locale related):
```
00000000001e0dec <_ZNK9BCollator11SetStrengthEa>:
  1e0dec: 41 11        	addi	sp, sp, -16
  1e0dee: 22 e0        	sd	s0, 0(sp)
  1e0df0: 06 e4        	sd	ra, 8(sp)
  1e0df2: 00 08        	addi	s0, sp, 16
  1e0df4: 7d 57        	li	a4, -1
  1e0df6: 63 80 e5 04  	beq	a1, a4, 0x1e0e36 <_ZNK9BCollator11SetStrengthEa+0x4a>
  1e0dfa: 11 47        	li	a4, 4
  1e0dfc: ae 87        	mv	a5, a1
  1e0dfe: 63 8e e5 02  	beq	a1, a4, 0x1e0e3a <_ZNK9BCollator11SetStrengthEa+0x4e>
  1e0e02: 63 43 b7 02  	blt	a4, a1, 0x1e0e28 <_ZNK9BCollator11SetStrengthEa+0x3c>
  1e0e06: 05 47        	li	a4, 1
  1e0e08: 81 45        	li	a1, 0
  1e0e0a: 63 86 e7 00  	beq	a5, a4, 0x1e0e16 <_ZNK9BCollator11SetStrengthEa+0x2a>
  1e0e0e: 09 47        	li	a4, 2
  1e0e10: 63 93 e7 02  	bne	a5, a4, 0x1e0e36 <_ZNK9BCollator11SetStrengthEa+0x4a>
  1e0e14: 85 45        	li	a1, 1
  1e0e16: 08 69        	ld	a0, 16(a0)
  1e0e18: 1c 61        	ld	a5, 0(a0)
  1e0e1a: dc 6f        	ld	a5, 152(a5) ; <-- crash here

```
Need to retest on clean Haiku system.

**Comment by LekKit** [comment]
I'm currently deciding some license-related changes, and I'd like to hear your feedback if possible (Could benefit UserlandVM)

I'm heavily against closed forks, together with @cerg2010cerg2010, but GPL is slightly restrictive sometimes when it comes to distributing/integrating the project. GPL **requires accompanying code to be GPL-compatible**, and **requires that modifications are always possible**.

I'm investigating the dual-licensing approach, where you aren't forced to open your code or change licensing when linking with upstream `librvvm` (no matter statically or dynamically), or distributing it. It would "just work" under terms of any other license if you are using upstream version, but if you want to modify `librvvm` sources you **should** make them available under GPL terms.

This usecase is perfectly covered by `Artistic License 2.0` which is also OSI-approved, `MPL` looks like a good choice as well. It would allow, for example, to distribute RVVM in Apple App Store (otherwise it would violate the **modifications are always possible** part of GPL).

I'm curious if this can help integrate RVVM with UserlandVM more easily, or allow distributing it in FreeBSD/Haiku repositories. Would like to hear your thoughts on that, or any other ideas you may have. I hope we can make a good compromise for both end users and us, developers)
After discussing this together with @cerg2010cerg2010 he approved this, as long as we don't relicense his code as overly permissive because closed forks are bad :D

**Comment by X547** [comment]
> I'm currently deciding some license-related changes, and I'd like to hear your feedback if possible (Could benefit UserlandVM)

librvvm license allowing to use it with MIT-licensed code would be beneficial for Haiku. It would allow to distribute RVVM based CPU engine plugin for UserlandVM as part of system. I originally planned to distribute TinyEMU based engine as default and RVVM engine as optionally installable.

**Comment by LekKit** [comment]
> > I'm currently deciding some license-related changes, and I'd like to hear your feedback if possible (Could benefit UserlandVM)
> 
> librvvm license allowing to use it with MIT-licensed code would be beneficial for Haiku. It would allow to distribute RVVM based CPU engine plugin for UserlandVM as part of system. I originally planned to distribute TinyEMU based engine as default and RVVM engine as optionally installable.

Will MPL or Artistic license be compatible with MIT? That looks to me like a nice middle-ground between permissive and copyleft, since we just drop the requirement that "modifications are always possible" and the viral nature of GPL linkage, while still insisting that changes made to `librvvm` code should be public. You may even build a proprietary version of UserlandVM which links to `librvvm` statically, and the only concern would be that you may no longer make modifications to `librvvm` internals and should use it as-is (Or publish your changes to it, without the need to relicense or open any other parts of your software).

**Comment by X547** [comment]
If it will be possible to link closed-sourced fork of UserlandVM to unmodified librvvm.a, it should be fine.

**Comment by LekKit** [comment]
> If it will be possible to link closed-sourced fork of UserlandVM to unmodified librvvm.a, it should be fine.

Hope that's resolved now, see the repo
The sources files still carry baggage of GPL headers, but the main repo licenses take precedence. They will be gradually migrated to SPDX or dual-headers, all the other contributors personally agreed with such changes (I hope you too, since I2C hid and Haiku window are copyrighted by you).
Discussions to further improve the situation are welcome. See 9cedfdf commit explanation for more details.

**Comment by LekKit** [comment]
> I found one regression with new `librvvm`: menu windows are displayed at wrong position and size. Issue is not present with old embedded RVVM code and with TinyEMU code. Probably FPU related because `float` type is used in GUI coordinates.

Depending on what reference commit you used in your fork, it could've lacked NaN boxing (Which is now implemented properly, see `src/mem_ops.h`), which fills upper 64-bit register part with `0xFFFFFFFF` when working with 32-bit floats. There are no other significant changes that could cause that.

I will be returning to my debugging efforts with legacy `librvvm` soon, but I want to upstream a few things in the meantime (RTL8169, Initial Linux user emulation support).

**Comment by LekKit** [comment]
Upstreamed a new reusable ELF loader lib. It loads userspace Linux ELFs into process address space fine, and even runs some of them (Syscall emulation and stack setup are unstable and limited, not upstreamed yet).
It also loads firmware/kernel ELFs now, same as QEMU does (basically in-house `objcopy -O binary`)

No FPU issues with rvvm-user + Linux binaries that I see so far. To be continued...

**Comment by LekKit** [comment]
New RVJIT optimization allows 1:1 translation of load/store instructions to host instructions in userland emulation, this should boost perf a **lot**, @X547 can you evaluate if it works OK with UserlandVM?

**Comment by fish4terrisa-MSDSM** [comment]
I am very interested in rvvm-user, can i build and try it on my machine? 
And, since currently rvvm VM mode is very fast, why not use 9pfs in the VM mode and run the binary in the guest instead?(Opensbi and linux kernel won't use too much memory). User mode will cause a lot of problems on premissions, syscalls, rootfs and many others.(See `proot` and `qemu-user`) That will also help if users want to transfer files between guest and host. (Maybe we can have a fake metadata mode for the 9pfs option, just like `humfs` in User mode linux, so that a normal user can boot the system from 9pfs)

**Comment by LekKit** [comment]
> I am very interested in rvvm-user, can i build and try it on my machine?

See cdfea8b. It is very WIP, and the entire point of userland emulator in future is that **more** perf can be squeezed, it fully integrates with your system (GUI goes directly over host X/Wayland, same with sound, etc) and even things like GPU drivers could work I guess. It is also very nice for foreign-arch build systems.

**Comment by LekKit** [comment]
Tiny progress on running dyn RISC-V Linux ELFs.. (Here running buildroot chroot & PrBoom)
Seems like argv alignment isn't done properly, different arg count may cause a crash

![image](https://github.com/LekKit/RVVM/assets/50500857/9e796860-8a32-4cd0-9608-f0fd4647033d)

**Comment by LekKit** [comment]
@X547 should I actually omit PC increment entirely for userland API (like TEMU does)? I think it's actually a good idea now, and if it should be done it better be early than late

Basically I figured the PC increments on faulty instructions (load/store traps, illegal instructions etc) as well which is not nice at all, whereas only incrementing PC after ECALL is also weird

**Comment by X547** [comment]
> @X547 should I actually omit PC increment entirely for userland API (like TEMU does)? I think it's actually a good idea now, and if it should be done it better be early than late

I am fine with both, but it would be nice to be consistent with how real RISC-V works (do not increment PC on ECALL).

**Comment by LekKit** [comment]
> I am fine with both, but it would be nice to be consistent with how real RISC-V works (do not increment PC on ECALL).

I will be doing exactly that soon, then. It will cause a slight breakage for UserlandVM tho - you will need to patch it to add 4 bytes to PC on `ecall`. Considering I am only doing this for v0.6 and `RVVM_ABI_VERSION` is different there, I guess it's ok.

UPD: cc3466c

**Comment by LekKit** [comment]
Many static binaries & Musl-linked dynamic binaries work in rvvm-user at this point.

However, there is a strange issue that glibc dyn ELFs do crash early in `ld-linux` when running rvvm-user in chroot on x86 host, yet work fine when running rvvm-user on a RISC-V host (I.e. emulating RISC-V userland on RISC-V).

I've implemented detailed guest exception handling and debug output, but struggle to debug this. If anyone is interested in joining the efforts, here is a direction:
- Get a 64-bit Linux host
- Uncomment this: https://github.com/LekKit/RVVM/blob/f24143719f09cd5532a15c491990072239e91b30/src/rvvm_user.c#L47
- Patch the beginning of `main()` in `main.c` to have this:
```c
int rvvm_user(int argc, char** argv, char** envp);

int main(int argc, char** argv, char** envp)
{
    if (argc > 2 && rvvm_strcmp(argv[1], "-user")) {
        rvvm_set_loglevel(LOG_INFO);
        return rvvm_user(argc - 2, argv + 2, envp);
    }
    ...
```
- Build RVVM with `make CFLAGS="-static"`
- Try running some RISC-V binaries with `rvvm -user [command]`

**Comment by X547** [comment]
> Try running some RISC-V binaries with `rvvm -user [command]`

I think that it would be better to put userland Linux emulator into separate executable, something like `rvvm-user-linux [command]`. Unlike machine emulation, userland emulator is very platform-dependent and non-portable.

**Comment by LekKit** [comment]
> I think that it would be better to put userland Linux emulator into separate executable, something like `rvvm-user-linux [command]`. Unlike machine emulation, userland emulator is very platform-dependent and non-portable.

It's a good idea, esp. considering that `binfmt-misc` only allows to pass argv to emulator. I just want some quick solution for now that doesn't need to modify ton of stuff in repo like build systems etc.

Honestly tho I like the idea that both user and machine modes share like 95% of emulation/JIT code. It's not entirely nonportable either (Its in theory possible to run Linux programs on Haiku or FreeBSD).

**Comment by X547** [comment]
> Honestly tho I like the idea that both user and machine modes share like 95% of emulation/JIT code.

This is what dynamic libraries for.

> It's not entirely nonportable either (Its in theory possible to run Linux programs on Haiku or FreeBSD).

So are you planning to implement Linux syscalls on top of POSIX, not like UserlandVM where host and guest syscalls are assumed to be the same? It would be much more complex and likely cause a lot of compatibility problems.

**Comment by LekKit** [comment]
> So are you planning to implement Linux syscalls on top of POSIX, not like UserlandVM where host and guest syscalls are assumed to be the same? It would be much more complex and likely cause a lot of compatibility problems.

So far I am simply calling POSIX/Linux specific libc functions that correspond to their syscall and it's not problematic. There are a few things that can be done with this to ensure maximum compatibility, like converting ABI/endianness in structs passed between kernel<->user, like QEMU does.

**Comment by LekKit** [comment]
@X547 big news: (TLDR: I got glibc dyn binaries to work)

We can't simply jump from guest syscalls to host kernel (as you suggested earlier). It turns out structures passed in Linux syscalls are defined subtly different on different architectures, so simply passing a pointer to `struct stat` from RISC-V process to x86_64 kernel results in corruption. So when I implemented conversion most of the mad crashes went away.

This also means that subtle difference between POSIX systems like Haiku or BSD can be also eliminated this way and allow to run some apps on a foreign kernel

**Comment by LekKit** [comment]
Managed to launch `pacman` in chroot. Some structs are not yet converted properly, and there are bugs, so it still doesn't work perfectly but at least something works x)

![image](https://github.com/LekKit/RVVM/assets/50500857/132e5bbc-7aa6-4b16-a7ca-363faa3d3359)

**Comment by X547** [comment]
> thread failed to start

I suppose some special handling is needed when spawning new threads like in UserlandVM.

**Comment by LekKit** [comment]
> I suppose some special handling is needed when spawning new threads like in UserlandVM.

I guess so. The struct conversion is still WIP so I'd need to finish it first, also I don't see that guest touch `clone()` syscall at all.

**Comment by LekKit** [comment]
Went further and now crashes immediately in the newly `clone()`'d thread. The main thread survives tho, and complains a lot.

![image](https://github.com/LekKit/RVVM/assets/50500857/531497e9-3f28-4e9c-9456-7cadd03eb942)

**Comment by LekKit** [comment]
Another useful trick: On a RISC-V host, selectively bypass syscalls directly to `syscall(NR, ...)` to bisect whether your shim is working properly.

There are only a few syscalls that won't ever work that way, one notable exception is `clone()` which, if executed directly would make a child thread return into RVVM emulation code and crash. Also this helped me to figure out that my `clone()` implementation is faulty because pacman is broken the same way even when all other syscalls are bypassed

**Comment by fish4terrisa-MSDSM** [comment]
It seems that currently RVVM have fake chroot and fake bind options. Maybe fake pid and uid can also be added? We may add a fake pid namespace to RVVM so that we can boot the system by running `rvvm-user /sbin/init` and get full service functions of systemd.

**Comment by LekKit** [comment]
> It seems that currently RVVM have fake chroot and fake bind options. Maybe fake pid and uid can also be added? We may add a fake pid namespace to RVVM so that we can boot the system by running `rvvm-user /sbin/init` and get full service functions of systemd.

Fake uid indeed could be implemented as in `proot` so the guest thinks it's running as root. However running systemd like that might turn out complicated, we'll see.

**Comment by LekKit** [comment]
I have reworked CMake build system so static libs can be built properly and not split into `librvvm.a`/`librvvm_cpu32.a`/`librvvm_cpu64.a`/`librvjit.a` - instead, only `librvvm_static.a` is provided. Shared library `librvvm.so` is also built by default now.

Building `rvvm` CLI binary is unchanged.

Perhaps I can also implement an `install` target if you need it for Haiku RVVM package/port. But `make install` does this job already for RVVM packaging purposes and CMake in this repo is more like a tool oriented for devs that are used to CMake, or use an OS that is incompatible with POSIX & Make (Windows).
